chanfitness/rate_limit.go

```diff
@@ -35,3 +49,27 @@ func getRateLimit(flapCount int) time.Duration {
 
 	return rateLimits[tier]
 }
+
+// cooldownFlapCount takes a timestamped flap count, and returns its value
+// scaled down by our cooldown factor if at least our cooldown period has
+// elapsed since the peer last flapped. We do this because we store all-time
+// flap count for peers, and want to allow downgrading of peers that have not
+// flapped for a long time.
+func cooldownFlapCount(now time.Time, flapCount int,
+	lastFlap time.Time) int {
+
+	// Calculate time since our last flap, and the number of times we need
+	// to apply our cooldown factor.
+	timeSinceFlap := now.Sub(lastFlap)
+	cooldownPeriods := timeSinceFlap / flapCountCooldownPeriod
```

Comment:
This way a flap of 100 would cool down to 1 in about 90 weeks.
We could be faster after the initial period. Also we could be more gradual if you allow fractional `coolDownPeriods` (just make `timeSinceFlap` float64).